### PR TITLE
feat: enable tree shaking - FE-4253

### PR DESCRIPTION
**Pull request by mkrds**
### Proposed behaviour

This PR changes the build process so that ES6 modules are now also provided along the old CommonJS ones

### Checklist

- [x] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [ ] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

1. Find a Sage repo that uses most recent version of `carbon-react` (sbc.accounting for instance)
2. Run build script from this repo (usually npm run build) and check the size of the built package (should be displayed in the terminal after succesful build)
3. After previous step install version of `carbon-react` based on build from this PR (there are instructions in the comment left by codesandbox bot, click second link from the top)
4. Run build again and compare the new size with the previous one (new one should be smaller)
5. Run unit tests in that repo to confirm that everything works.

### CODE REVIEW

**Review by samtjo — DISMISSED** [review]

File: docs/bundle-size.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,36 @@
+<Meta title="Documentation/Bundle Size" />
+
+# Bundle size
+
+How to minimize bundle size
+
+## Tree shaking
+
+Tree shaking in Carbon should work out of the box for most modern frameworks.
+As long as you're using a bundler that supports ES6 modules and tree shaking there's no need to change the import paths of the components.
```
Comment (jamime) [inline docs/bundle-size.stories.mdx:10]:
```suggestion
Tree shaking in Carbon should work out of the box as long as you're using a bundler that supports ES6 modules and tree shaking. There is no need to change the import paths of the components.
```

File: docs/bundle-size.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+```jsx
+import Component from "carbon-react/lib/components/component";
+import {
+  SubComponent1,
+  SubComponent2,
+} from "carbon-react/lib/components/other-component";
+```
+
+To make sure that your webpack config supports tree shaking ensure that none of these optimization options is turned off.
```
Comment (jamime) [inline docs/bundle-size.stories.mdx:21]:
```suggestion
To make sure that your webpack config supports tree shaking ensure that none of these optimization options are turned off.
```

File: docs/bundle-size.stories.mdx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+These are set to true by default in the recent webpack versions so you probably won't have to do any changes in your webpack config.
```
Comment (jamime) [inline docs/bundle-size.stories.mdx:22]:
```suggestion
These values are set to `true` by default.
```

File: docs/bundle-size.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  //...
+  optimization: {
+    usedExports: true,
+    sideEffects: true,
+    innerGraph: true,
+  },
+};
+```
+
+For more information about the topic see webpack docs - https://webpack.js.org/guides/tree-shaking/
```
Comment (jamime) [inline docs/bundle-size.stories.mdx:36]:
```suggestion
For more information about the topic see the [webpack guide to tree shaking](https://webpack.js.org/guides/tree-shaking/).
```

**Review by jamime — CHANGES_REQUESTED** [review]

**Review by jamime — DISMISSED** [review]

**Review by samtjo — APPROVED** [review]

File: scripts/build.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    "--extensions",
+    `"${extensions.join(",")}"`,
+    srcDir,
+    "--out-dir",
+    outDir,
+    "--ignore",
+    `"${ignore.join('","')}"`,
+  ];
+
+  const command = ["cross-env babel", ...babelArgs].join(" ");
```
Comment (samtjo) [inline scripts/build.js:48]: I don't think we need `cross-env` here unless we need to add `cross-env NODE_ENV=production` in (Which I don't think we do as it's being set below)

**Review by samtjo — COMMENTED** [review]
Looks good, just a couple of small questions/comments from me

File: scripts/build.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    "**/*.test.tsx",
+    "**/*.spec.ts",
+    "**/*.spec.tsx",
+    "**/*.d.ts",
+  ];
+
+  const outDir = path.resolve(
+    relativeOutDir,
+    {
+      node: "./",
```
Comment (jamime) [inline scripts/build.js:31]:
```suggestion
      cjs: "./",
```

File: scripts/build.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      stable: "./esm",
```
Comment (jamime) [inline scripts/build.js:32]:
```suggestion
      esm: "../esm",
```

**Review by jamime — CHANGES_REQUESTED** [review]

File: scripts/build.js (hunk 5 above)
Comment (mkrds) [reply scripts/build.js:48]: I actually got some errors when I simply wanted to run `npm run babel` thats why I decided to stick with `cross-env`

**Review by samtjo — DISMISSED** [review]

**Review by grabkowski — APPROVED** [review]

**Review by grabkowski — APPROVED** [review]

**Review by samtjo — APPROVED** [review]

**Review by samtjo — DISMISSED** [review]

**Review by samtjo — APPROVED** [review]

**Review by chrisbarber86 — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Pull request merged**